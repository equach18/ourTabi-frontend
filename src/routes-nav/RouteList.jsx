import { useContext } from "react";
import { Routes, Route, Navigate } from "react-router-dom";
import Home from "../pages/Home";
import Dashboard from "../pages/Dashboard";
import LoginForm from "../components/user/LoginForm";
import SignupForm from "../components/user/SignupForm";
import NewTripForm from "../components/trip/NewTripForm";
import TripDetails from "../pages/TripDetails";
import NewActivityForm from "../components/activity/NewActivityForm";
import NotFound from "../pages/NotFound";
import PrivateRoute from "./PrivateRoute";
import UserContext from "../context/UserContext";
import TripProvider from "../context/TripProvider";
import EditTripForm from "../components/trip/EditTripForm";
import EditActivityForm from "../components/activity/EditActivityForm";
import EditProfileForm from "../components/user/EditProfileForm";
import TripSearchResults from "../pages/TripSearchResults";

function RouteList() {
  const {
    currentUser,
    trips,
    login,
    signup,
    addTrip,
    editTrip,
    isLoading,
    removeTrip,
    updateUser,
  } = useContext(UserContext);

  return (
    <Routes>
      <Route path="/" element={<Home currentUser={currentUser} />} />
      <Route path="/login" element={<LoginForm login={login} />} />
      <Route path="/signup" element={<SignupForm signup={signup} />} />

      {/* Protected Routes: Only accessible if logged in */}
      <Route
        element={
          <PrivateRoute currentUser={currentUser} isLoading={isLoading} />
        }
      >
        <Route
          path="/dashboard"
          element={<Dashboard currentUser={currentUser} trips={trips} />}
        />
        <Route
          path="/profile/edit"
          element={
            <EditProfileForm
              currentUser={currentUser}
              updateUser={updateUser}
            />
          }
        />
        <Route path="/trips/new" element={<NewTripForm addTrip={addTrip} />} />
        <Route
          path="/trips/:tripId"
          element={
            <TripProvider>
              <TripDetails currentUser={currentUser} removeTrip={removeTrip} />
            </TripProvider>
          }
        />
        <Route
          path="/trips/:tripId/edit"
          element={<EditTripForm editTrip={editTrip} trips={trips} />}
        />
        <Route path="/search" element={<TripSearchResults />} />
        <Route
          path="/trips/:tripId/activities/new"
          element={
            <TripProvider>
              <NewActivityForm />
            </TripProvider>
          }
        />
        <Route
          path="/trips/:tripId/activities/:activityId/edit"
          element={
            <TripProvider>
              <EditActivityForm />
            </TripProvider>
          }
        />
      </Route>
      {/* Handle 404 Not Found */}
      <Route path="*" element={<NotFound />} />
    </Routes>
  );
}

export default RouteList;
