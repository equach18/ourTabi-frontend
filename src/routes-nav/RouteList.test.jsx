import { screen } from "@testing-library/react";
import RouteList from "./RouteList";
import {
  renderWithAllProviders,
  testUser1,
  testTrip,
  defaultUserContext,
} from "../test-helpers/_testCommons";

// mock components
vi.mock("../pages/Home", () => ({
  default: () => <div>Home Page</div>,
}));
vi.mock("../components/user/LoginForm", () => ({
  default: () => <div>Login Form</div>,
}));
vi.mock("../components/user/SignupForm", () => ({
  default: () => <div>Signup Form</div>,
}));
vi.mock("../pages/Dashboard", () => ({
  default: () => <div>Dashboard</div>,
}));
vi.mock("../pages/NotFound", () => ({
  default: () => <div>Not Found</div>,
}));

describe("RouteList", () => {
  it("matches snapshot for home route", () => {
    const { asFragment } = renderWithAllProviders(<RouteList />, {
      route: "/",
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for login route", () => {
    const { asFragment } = renderWithAllProviders(<RouteList />, {
      route: "/login",
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for unmatched route", () => {
    const { asFragment } = renderWithAllProviders(<RouteList />, {
      route: "/no-such-route",
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for dashboard when logged in", () => {
    const { asFragment } = renderWithAllProviders(<RouteList />, {
      route: "/dashboard",
      userContext: {
        currentUser: { id: 1, username: "testuser" },
        trips: [],
        isLoading: false,
      },
    });
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders public routes correctly", () => {
    renderWithAllProviders(<RouteList />, {
      route: "/login",
    });
    expect(screen.getByText("Login Form")).toBeInTheDocument();

    renderWithAllProviders(<RouteList />, { route: "/" });
    expect(screen.getByText("Home Page")).toBeInTheDocument();

    renderWithAllProviders(<RouteList />, { route: "/signup" });
    expect(screen.getByText("Signup Form")).toBeInTheDocument();
  });

  it("redirects to login if not logged in on protected routes", () => {
    renderWithAllProviders(<RouteList />, {
      route: "/dashboard",
      userContext: {
        currentUser: null,
        login: vi.fn(),
      },
    });

    expect(screen.getByText("Login Form")).toBeInTheDocument();
  });

  it("shows protected page if logged in", () => {
    const userContext = {
      ...defaultUserContext,
      currentUser: testUser1,
      trips: [testTrip],
      isLoading: false,
    };

    renderWithAllProviders(<RouteList />, {
      route: "/dashboard",
      userContext,
    });

    expect(screen.getByText("Dashboard")).toBeInTheDocument();
  });

  it("shows 404 page for unmatched route", () => {
    renderWithAllProviders(<RouteList />, {
      route: "/badRoute",
    });
    expect(screen.getByText("Not Found")).toBeInTheDocument();
  });
});
