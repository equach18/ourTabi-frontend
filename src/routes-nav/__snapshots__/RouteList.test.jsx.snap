// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RouteList > matches snapshot for dashboard when logged in 1`] = `
<DocumentFragment>
  <div>
    Dashboard
  </div>
</DocumentFragment>
`;

exports[`RouteList > matches snapshot for home route 1`] = `
<DocumentFragment>
  <div>
    Home Page
  </div>
</DocumentFragment>
`;

exports[`RouteList > matches snapshot for login route 1`] = `
<DocumentFragment>
  <div>
    Login Form
  </div>
</DocumentFragment>
`;

exports[`RouteList > matches snapshot for unmatched route 1`] = `
<DocumentFragment>
  <div>
    Not Found
  </div>
</DocumentFragment>
`;
